File: src/runregistry.ts
import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

/**
 * Run dedup / latest-wins (Sweep lesson): one active pipeline per ticket key
 * across processes. Lock files under <home>/locks/<sanitized-ticket>.lock;
 * a stale lock (older than ttlMs) is broken rather than blocking forever.
 */

export interface RunLock {
  ticketId: string;
  path: string;
  release(): void;
}

export function sanitizeKey(ticketId: string): string {
  return ticketId.replace(/[^A-Za-z0-9._-]/g, '_');
}

export function tryAcquireRun(
  homeDir: string,
  ticketId: string,
  opts: { ttlMs?: number; now?: () => number } = {},
): RunLock | null {
  const locksDir = join(homeDir, 'locks');
  mkdirSync(locksDir, { recursive: true });
  const path = join(locksDir, `${sanitizeKey(ticketId)}.lock`);
  const now = (opts.now ?? Date.now)();
  const ttl = opts.ttlMs ?? 60 * 60_000;

  if (existsSync(path)) {
    let holder: { startedAt?: number } | null = null;
    try {
      holder = JSON.parse(readFileSync(path, 'utf8')) as { startedAt?: number };
    } catch {
      holder = null;
    }
    const age = now - (holder?.startedAt ?? 0);
    if (holder && age <= ttl) return null; // someone else holds it fresh
    // stale or corrupt: break the lock (latest-wins)
    rmSync(path, { force: true });
  }

  writeFileSync(path, JSON.stringify({ pid: process.pid, startedAt: now }));
  let released = false;
  return {
    ticketId,
    path,
    release: () => {
      if (released) return;
      released = true;
      rmSync(path, { force: true });
    },
  };
}

File: internal/pipeline/runregistry.go
// Run dedup / latest-wins: the Go port of src/runregistry.ts (sweep lesson).
// One active pipeline per ticket key across processes. Lock files live under
// <home>/locks/<sanitized-ticket>.lock; a stale lock (older than the one-hour
// TTL) is broken rather than blocking forever.
//
// The canonical port lives in internal/ledger next to the run logger that
// shares its storage layout; this thin wrapper keeps the pinned
// pipeline.TryAcquireRun(home, ticketID) shape for the cli wiring, where
// home is the resolved DEVAGENT_HOME ("$DEVAGENT_HOME || ~/.devagent").

package pipeline

import "github.com/FreePeak/devagent/internal/ledger"

// RunLock mirrors runregistry.ts RunLock. Release is safe to call twice.
type RunLock = ledger.RunLock

// TryAcquireRun mirrors tryAcquireRun: acquire the run lock for ticketID, or
// nil when a fresh lock is held elsewhere. A stale or corrupt lock is broken
// (latest-wins) and the payload is {"pid":<pid>,"startedAt":<ms>}.
func TryAcquireRun(home, ticketID string) *RunLock {
	return ledger.TryAcquireRun(home, ticketID, 0)
}
